### ISSUE TITLE
Salt Master Hang

### ISSUE THREAD

**Issue by jakehilton** [saltstack/salt#29674]
I've got 8 salt masters running in pairs for redundancy. So 4 groups handling roughly 500-600 servers each. Every few days the masters will hang. They both hang at the same time and requires a restart to help clear things up.

I have the minions run a highstate every 30 min. If I used the documented highstate schedule (https://docs.saltstack.com/en/latest/topics/jobs/schedule.html#highstates) things would blow up so I'm using a simple schedule cmd.run. I'm also running a cleanup script from the masters every 30 minutes to clean out stale minions.

The command I'm running to clear out stale minions is:

```
salt --out txt "*" test.ping | grep "Not connected" | cut -d ":" -f 1 | xargs -I dead_minion salt-key -y -d dead_minion
```

I did test out `salt-run manage.down removekeys=True` but it didn't seem to work as well.

I'm not seeing any issues with minions being removed per se.. but I am seeing salt master hanging. So when new boxes come online they aren't provisioned properly.

I"m seeing errors in the that seem to suggest that an internal minion check is failing miserably. The log is below.

A log section:
```
2015-12-13 06:53:53,199 [salt.master      ][ERROR   ][17378] Error in function _mine:
Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/salt/master.py", line 1379, in run_func
    ret = getattr(self, func)(load)
  File "/usr/lib/python2.7/dist-packages/salt/master.py", line 1007, in _mine
    load = self.__verify_load(load, ('id', 'data', 'tok'))
  File "/usr/lib/python2.7/dist-packages/salt/master.py", line 924, in __verify_load
    if not self.__verify_minion(load['id'], load['tok']):
  File "/usr/lib/python2.7/dist-packages/salt/master.py", line 824, in __verify_minion
    with salt.utils.fopen(pub_path, 'r') as fp_:
  File "/usr/lib/python2.7/dist-packages/salt/utils/__init__.py", line 1204, in fopen
    fhandle = open(*args, **kwargs)
IOError: [Errno 2] No such file or directory: '/etc/salt/pki/master/minions/ause1-encoder546'
2015-12-13 06:54:25,104 [salt.transport.zeromq][ERROR   ][17378] Bad load from minion
2015-12-13 06:54:35,228 [salt.transport.zeromq][ERROR   ][17370] Bad load from minion
2015-12-13 06:54:46,744 [salt.transport.zeromq][ERROR   ][17398] Bad load from minion
2015-12-13 06:54:53,352 [salt.master      ][ERROR   ][17317] Error in function _mine:
Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/salt/master.py", line 1379, in run_func
    ret = getattr(self, func)(load)
  File "/usr/lib/python2.7/dist-packages/salt/master.py", line 1007, in _mine
    load = self.__verify_load(load, ('id', 'data', 'tok'))
  File "/usr/lib/python2.7/dist-packages/salt/master.py", line 924, in __verify_load
    if not self.__verify_minion(load['id'], load['tok']):
  File "/usr/lib/python2.7/dist-packages/salt/master.py", line 824, in __verify_minion
    with salt.utils.fopen(pub_path, 'r') as fp_:
  File "/usr/lib/python2.7/dist-packages/salt/utils/__init__.py", line 1204, in fopen
    fhandle = open(*args, **kwargs)
IOError: [Errno 2] No such file or directory: '/etc/salt/pki/master/minions/ause1-encoder546'
2015-12-13 06:55:00,814 [salt.master      ][ERROR   ][17343] Error in function _mine:
Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/salt/master.py", line 1379, in run_func
    ret = getattr(self, func)(load)
  File "/usr/lib/python2.7/dist-packages/salt/master.py", line 1007, in _mine
    load = self.__verify_load(load, ('id', 'data', 'tok'))
  File "/usr/lib/python2.7/dist-packages/salt/master.py", line 924, in __verify_load
    if not self.__verify_minion(load['id'], load['tok']):
  File "/usr/lib/python2.7/dist-packages/salt/master.py", line 824, in __verify_minion
    with salt.utils.fopen(pub_path, 'r') as fp_:
  File "/usr/lib/python2.7/dist-packages/salt/utils/__init__.py", line 1204, in fopen
    fhandle = open(*args, **kwargs)
IOError: [Errno 2] No such file or directory: '/etc/salt/pki/master/minions/ause1-encoder646'
2015-12-13 06:55:17,444 [salt.master      ][ERROR   ][17443] Error in function _mine:
Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/salt/master.py", line 1379, in run_func
    ret = getattr(self, func)(load)
  File "/usr/lib/python2.7/dist-packages/salt/master.py", line 1007, in _mine
    load = self.__verify_load(load, ('id', 'data', 'tok'))
  File "/usr/lib/python2.7/dist-packages/salt/master.py", line 924, in __verify_load
    if not self.__verify_minion(load['id'], load['tok']):
  File "/usr/lib/python2.7/dist-packages/salt/master.py", line 824, in __verify_minion
    with salt.utils.fopen(pub_path, 'r') as fp_:
  File "/usr/lib/python2.7/dist-packages/salt/utils/__init__.py", line 1204, in fopen
    fhandle = open(*args, **kwargs)
IOError: [Errno 2] No such file or directory: '/etc/salt/pki/master/minions/ause1-encoder66'
```

I'm also seeing endless logs about the "bad load from minion" even after I ping out to all the minions.

The multi-master setup I'm using is:
https://docs.saltstack.com/en/latest/topics/tutorials/multimaster.html

Would this be better?
https://docs.saltstack.com/en/latest/topics/tutorials/multimaster_pki.html

Ideally having two hot is great but if it causes issues then obviously I'll back off of it.

Any help in determining the core issue would be much appreciated.

Thanks!

**Comment by jakehilton** [comment]
I forgot to mention I'm running the latest salt.. 2015.8.3 on ubuntu 14.04.3

A little more information about minion removing as well.. For instance.. on one box with 561 minions I just tried
```
# salt-run manage.down removekeys=True
Salt request timed out. The master is not responding. If this error persists after verifying the master is up, worker_threads may need to be increased.
```

I've got my workers at 20 right now on that box.

On it's sister box with 50 worker threads configured I get the same error.

**Comment by jfindlay** [comment]
@jakehilton, thanks for the report.  @meggiebot, @cachedout, do you have any ideas here?

**Comment by Ch3LL** [comment]
@jakehilton can you please post any related multi- master configurations, particularly the minion configuration file. I want to attempt to try to recreate this issue but want to ensure I have the same multi-master settings. Also is there any other related information in the debug logs when the master hangs? 

In your logs it keeps stating it does not see the key for the 'ause1-encoder646' minion. Can you run `salt-key` and see if its in your list of minions please.  

Also I found this issue: #15471  Are you using salt-call anywhere while the salt-minion process is not running?

**Comment by jakehilton** [comment]
Thank you for looking into this issue.

My master config looks like so on both masters:

```
open_mode: True
top_file_merging_strategy: same

file_roots:
  base:
    - /etc/salt/git/salt

pillar_roots:
  base:
    - /etc/salt/git/salt/pillar

worker_threads: 30
timeout: 30
```

My minion config looks like so:
```
random_reauth_delay: 60
recon_default: 1000
recon_max: 5000
recon_randomize: True
keysize: 2048
startup_states: highstate
master:
  - smaster1.server.com
  - smaster2.server.com
grains:
  salt_setup: true
  ultype: smaster
  ulsite: uplynk
id: ause1-encoder646
```

There was really only that cleanup call that ran on a schedule that I think is causing the issue. What is strange is that it's trying to reference a minion file that is gone.. and it seemingly blows up because of that.

I do show that the minion is still listed.
```
# salt-key | grep 646
ause1-encoder646
```

It does list quite a few minion files.. I didn't send the whole log because it repeated the error for quite a few minions.

Thanks for your help in this!

**Comment by Ch3LL** [comment]
@jakehilton I have setup the following:

2 masters with the same configurations as yours
1 minion with same configurations as yours

I understand this is not the exact scenario as yours but suspect I should be able to recreate the same thing. I have tried running the `salt --out txt "*" test.ping | grep "Not connected" | cut -d ":" -f 1 | xargs -I dead_minion salt-key -y -d dead_minion` command and have not been able to recreate your issue. 

My other thought is the state.highstate schedule you have setup. Can you post your scheduler configurations for both the salt-key -d command and the highstate so I can try to exactly replicate your scenario.

My other suggestion would be to remove one of these out at a time to see if either the highstate or the salt-key delete commands are causing this behvaior. 

I will keep monitoring the servers today and see if I might see the salt-masters hang at a later time.

**Comment by jakehilton** [comment]
It did take a few days for the hang to occur. It was when I had both schedules running in concert.

Since removing the schedule to clear out dead minions the salt masters seem to have stabilized. I'm continuing to monitor it now.

My current schedule yaml looks like this
```
schedule:
  highstate:
    function: state.highstate
    minutes: 60
    maxrunning: 1
```

My problematic schedule yaml looked like this:
```
schedule:
  highstate:
    function: cmd.run
    minutes: 30
    args:
      - 'salt-call -l warning --log-file-level=warning state.highstate'
    kwargs:
      stateful: False
      shell: /bin/sh

  minion_cleanup:
    function: cmd.run
    minutes: 30
    args:
      - 'salt --out txt "*" test.ping | grep "Not connected" | cut -d ":" -f 1 | xargs -I dead_minion salt-key -y -d dead_minion'
    kwargs:
      stateful: False
      shell: /bin/sh
```

You may also need to have a lot of minions coming and going to reproduce this issue.. seeing as we have some churn with minions. Not sure if you'll see it without some level of turn over.

**Comment by Ch3LL** [comment]
@jakehilton okay i've setup the schedules with 3 minions and i'll take one minion out of the picture to test this. I'll wait a day or two to see if the master is hanging.

I just re-read your earlier response when you stated that you could see the key of the minion that was in error in the log when running salt-key. Is it an accepted minion? Is this minion suppose to be a deleted key?

are your salt-masters still stabilized since removing the schedule to remove the keys? I'm curious if the problem is related to these jobs running at the same time. I'm not entirely sure though and I am guessing someone else would have some more insight to why this might be occurring. 

@cachedout do you have any ideas as to why this is occuring?

In the meantime I will continue to see if i can recreate this issue and post my results probably in the next couple of days, since you stated it takes a few days for the master to hang.

**Comment by jakehilton** [comment]
You may need to introduce some minion churn.. minions coming and going. Not sure if that's possible with your test setup.

We use an open master so all keys are accepted automatically. The script is supposed to cleanup minion connections that are stale. So it wasn't a deleted key when this script ran I believe.

So far the salt masters have not exhibited the hang after removing that call to cleanup stale minions. I am going to do a full test on all the masters this morning to make sure everything is still happy. 

Thank you.

**Comment by jakehilton** [comment]
I just had two masters hung.. with a lot of these messages before the freeze:

```
2015-12-16 18:45:59,229 [salt.transport.zeromq][ERROR   ][10409] Bad load from minion
2015-12-16 18:45:59,230 [salt.transport.zeromq][ERROR   ][10471] Bad load from minion
2015-12-16 18:45:59,350 [salt.transport.zeromq][ERROR   ][10409] Bad load from minion
2015-12-16 18:45:59,504 [salt.transport.zeromq][ERROR   ][10442] Bad load from minion
```

That happened on one of the two multi-masters.. nothing else in the logs then just a ton of those.

**Comment by jakehilton** [comment]
Another round of errors.. and a master hang. I safely removed a minion using salt-key -d <minionid> but then it looks like later it tried to act on that for some reason.

```
2015-12-18 18:08:12,006 [salt.master      ][ERROR   ][1736] Received minion error from [ause2-services04]: The minion function caused an exception
2015-12-18 18:51:22,967 [salt.master      ][ERROR   ][1657] Received minion error from [ip-10-212-133-231.us-west-2.compute.internal]: The minion function caused an exception
2015-12-18 18:59:32,671 [salt.master      ][ERROR   ][1721] Received minion error from [ause4-services14]: The minion function caused an exception
2015-12-18 19:02:16,021 [salt.master      ][ERROR   ][1613] Error in function _mine:
Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/salt/master.py", line 1379, in run_func
    ret = getattr(self, func)(load)
  File "/usr/lib/python2.7/dist-packages/salt/master.py", line 1007, in _mine
    load = self.__verify_load(load, ('id', 'data', 'tok'))
  File "/usr/lib/python2.7/dist-packages/salt/master.py", line 924, in __verify_load
    if not self.__verify_minion(load['id'], load['tok']):
  File "/usr/lib/python2.7/dist-packages/salt/master.py", line 824, in __verify_minion
    with salt.utils.fopen(pub_path, 'r') as fp_:
  File "/usr/lib/python2.7/dist-packages/salt/utils/__init__.py", line 1204, in fopen
    fhandle = open(*args, **kwargs)
IOError: [Errno 2] No such file or directory: '/etc/salt/pki/master/minions/ause3-'
2015-12-18 19:04:29,552 [salt.master      ][ERROR   ][1714] Error in function _mine:
Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/salt/master.py", line 1379, in run_func
    ret = getattr(self, func)(load)
  File "/usr/lib/python2.7/dist-packages/salt/master.py", line 1007, in _mine
    load = self.__verify_load(load, ('id', 'data', 'tok'))
  File "/usr/lib/python2.7/dist-packages/salt/master.py", line 924, in __verify_load
    if not self.__verify_minion(load['id'], load['tok']):
  File "/usr/lib/python2.7/dist-packages/salt/master.py", line 824, in __verify_minion
    with salt.utils.fopen(pub_path, 'r') as fp_:
  File "/usr/lib/python2.7/dist-packages/salt/utils/__init__.py", line 1204, in fopen
    fhandle = open(*args, **kwargs)
IOError: [Errno 2] No such file or directory: '/etc/salt/pki/master/minions/ause3-'
```

**Comment by cachedout** [comment]
Hi @jakehilton Please see if #29962 resolves this problem for you.

**Comment by jakehilton** [comment]
Thank you I'll get this implemented and report back.

**Comment by cachedout** [comment]
@jakehilton It didn't pass the test suite. Hold off on trying to implement this until I can figure out what went wrong.

**Comment by jakehilton** [comment]
Roger that. Thanks!

**Comment by jakehilton** [comment]
I just got this error from a minion when running a highstate command from the master.

```
The minion function caused an exception: Traceback (most recent call last):
                    File "/usr/lib/python2.7/dist-packages/salt/minion.py", line 1039, in _thread_return
                      return_data = func(*args, **kwargs)
                    File "/usr/lib/python2.7/dist-packages/salt/modules/state.py", line 543, in highstate
                      st_ = salt.state.HighState(opts, pillar, kwargs.get('__pub_jid'), proxy=__proxy__)
                    File "/usr/lib/python2.7/dist-packages/salt/state.py", line 3170, in __init__
                      self.state = State(self.opts, pillar, jid, proxy=proxy)
                    File "/usr/lib/python2.7/dist-packages/salt/state.py", line 604, in __init__
                      self.opts['pillar'] = self._gather_pillar()
                    File "/usr/lib/python2.7/dist-packages/salt/state.py", line 626, in _gather_pillar
                      ret = pillar.compile_pillar()
                    File "/usr/lib/python2.7/dist-packages/salt/pillar/__init__.py", line 163, in compile_pillar
                      dictkey='pillar',
                    File "/usr/lib/python2.7/dist-packages/salt/utils/async.py", line 73, in wrap
                      ret = self._block_future(ret)
                    File "/usr/lib/python2.7/dist-packages/salt/utils/async.py", line 83, in _block_future
                      return future.result()
                    File "/usr/lib/python2.7/dist-packages/tornado/concurrent.py", line 215, in result
                      raise_exc_info(self._exc_info)
                    File "/usr/lib/python2.7/dist-packages/tornado/gen.py", line 879, in run
                      yielded = self.gen.send(value)
                    File "/usr/lib/python2.7/dist-packages/salt/transport/zeromq.py", line 160, in crypted_transfer_decode_dictentry
                      aes = cipher.decrypt(ret['key'])
                    File "/usr/lib/python2.7/dist-packages/Crypto/Cipher/PKCS1_OAEP.py", line 227, in decrypt
                      raise ValueError("Incorrect decryption.")
                  ValueError: Incorrect decryption.
```

This is with stock 2015.8.3 so nothing patched on it. Not sure if this would cause issues.. but thought I'd pass it along seeing as it does look to relate to the minion error above.

**Comment by jakehilton** [comment]
When I logged into the box that threw that exception from above.. I could run a salt-call state.highstate without error. Not sure how the call from the master would cause this error to pop.

**Comment by jakehilton** [comment]
Another error just popped.

```
The minion function caused an exception: Traceback (most recent call last):
                    File "/usr/lib/python2.7/dist-packages/salt/minion.py", line 1039, in _thread_return
                      return_data = func(*args, **kwargs)
                    File "/usr/lib/python2.7/dist-packages/salt/modules/state.py", line 543, in highstate
                      st_ = salt.state.HighState(opts, pillar, kwargs.get('__pub_jid'), proxy=__proxy__)
                    File "/usr/lib/python2.7/dist-packages/salt/state.py", line 3170, in __init__
                      self.state = State(self.opts, pillar, jid, proxy=proxy)
                    File "/usr/lib/python2.7/dist-packages/salt/state.py", line 604, in __init__
                      self.opts['pillar'] = self._gather_pillar()
                    File "/usr/lib/python2.7/dist-packages/salt/state.py", line 626, in _gather_pillar
                      ret = pillar.compile_pillar()
                    File "/usr/lib/python2.7/dist-packages/salt/pillar/__init__.py", line 163, in compile_pillar
                      dictkey='pillar',
                    File "/usr/lib/python2.7/dist-packages/salt/utils/async.py", line 73, in wrap
                      ret = self._block_future(ret)
                    File "/usr/lib/python2.7/dist-packages/salt/utils/async.py", line 83, in _block_future
                      return future.result()
                    File "/usr/lib/python2.7/dist-packages/tornado/concurrent.py", line 215, in result
                      raise_exc_info(self._exc_info)
                    File "/usr/lib/python2.7/dist-packages/tornado/gen.py", line 879, in run
                      yielded = self.gen.send(value)
                    File "/usr/lib/python2.7/dist-packages/salt/transport/zeromq.py", line 160, in crypted_transfer_decode_dictentry
                      aes = cipher.decrypt(ret['key'])
                    File "/usr/lib/python2.7/dist-packages/Crypto/Cipher/PKCS1_OAEP.py", line 197, in decrypt
                      m = self._key.decrypt(ct)
                    File "/usr/lib/python2.7/dist-packages/Crypto/PublicKey/RSA.py", line 174, in decrypt
                      return pubkey.pubkey.decrypt(self, ciphertext)
                    File "/usr/lib/python2.7/dist-packages/Crypto/PublicKey/pubkey.py", line 93, in decrypt
                      plaintext=self._decrypt(ciphertext)
                    File "/usr/lib/python2.7/dist-packages/Crypto/PublicKey/RSA.py", line 237, in _decrypt
                      cp = self.key._blind(ciphertext, r)
                  ValueError: Message too large
```

**Comment by jakehilton** [comment]
Had another hang.. it seems the master is iterating over files from removed minions..

```
IOError: [Errno 2] No such file or directory: '/etc/salt/pki/master/minions/ause1-encoder316'
2016-01-02 06:14:11,011 [salt.master      ][ERROR   ][9944] Error in function _mine:
Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/salt/master.py", line 1379, in run_func
    ret = getattr(self, func)(load)
  File "/usr/lib/python2.7/dist-packages/salt/master.py", line 1007, in _mine
    load = self.__verify_load(load, ('id', 'data', 'tok'))
  File "/usr/lib/python2.7/dist-packages/salt/master.py", line 924, in __verify_load
    if not self.__verify_minion(load['id'], load['tok']):
  File "/usr/lib/python2.7/dist-packages/salt/master.py", line 824, in __verify_minion
    with salt.utils.fopen(pub_path, 'r') as fp_:
  File "/usr/lib/python2.7/dist-packages/salt/utils/__init__.py", line 1204, in fopen
    fhandle = open(*args, **kwargs)
IOError: [Errno 2] No such file or directory: '/etc/salt/pki/master/minions/ause2-content51'
2016-01-02 06:14:11,315 [salt.master      ][ERROR   ][9964] Error in function _mine:
Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/salt/master.py", line 1379, in run_func
    ret = getattr(self, func)(load)
  File "/usr/lib/python2.7/dist-packages/salt/master.py", line 1007, in _mine
    load = self.__verify_load(load, ('id', 'data', 'tok'))
  File "/usr/lib/python2.7/dist-packages/salt/master.py", line 924, in __verify_load
    if not self.__verify_minion(load['id'], load['tok']):
  File "/usr/lib/python2.7/dist-packages/salt/master.py", line 824, in __verify_minion
    with salt.utils.fopen(pub_path, 'r') as fp_:
  File "/usr/lib/python2.7/dist-packages/salt/utils/__init__.py", line 1204, in fopen
    fhandle = open(*args, **kwargs)
IOError: [Errno 2] No such file or directory: '/etc/salt/pki/master/minions/ause1-encoder430'
```

I had removed them this time using `salt-run manage.down removekeys=True`

**Comment by cachedout** [comment]
*Very* interesting. This is a definite clue. I'll try to come up with a test case for this. Great work!

**Comment by jakehilton** [comment]
Just saw another few exceptions.. 
```
2016-01-04 04:04:37,512 [salt.loaded.int.returner.local_cache][ERROR   ][17796] prep_jid could not store a jid after 5 tries.
2016-01-04 04:04:37,512 [salt.master      ][ERROR   ][17796] Could not store job information for load: {'tgt_type': 'glob', 'jid': 'req', 'return': None, 'tgt': 'ausw2-encoder158', 'schedule': '__master_alive', 'cmd': '_return', 'pid': 24200, 'arg': [], 'fun': 'status.master', 'id': 'ausw2-encoder158'}
2016-01-04 04:04:40,832 [tornado.application][ERROR   ][17796] Future exception was never retrieved: Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/tornado/gen.py", line 230, in wrapper
    yielded = next(result)
  File "/usr/lib/python2.7/dist-packages/salt/transport/zeromq.py", line 530, in handle_message
    stream.send(self.serial.dumps(self._auth(payload['load'])))
  File "/usr/lib/python2.7/dist-packages/salt/transport/mixins/auth.py", line 383, in _auth
    with salt.utils.fopen(pubfn, 'w+') as fp_:
  File "/usr/lib/python2.7/dist-packages/salt/utils/__init__.py", line 1204, in fopen
    fhandle = open(*args, **kwargs)
IOError: [Errno 28] No space left on device: '/etc/salt/pki/master/minions/new-minion'
```

So not sure why it couldn't store a jid.. and then I see an error about no space left on the device. I checked and the device has ample space on it.. so not sure why it would be complaining about that. There are tons of errors about not being able to store the jid.. but didn't want to clutter this up with those.

These are from a master that has a bit of churn.. minions coming and going regularly.. and some of the minion ids may be re-used in some cases.. not sure if that helps.

**Comment by cachedout** [comment]
Did that device run out of inodes?

**Comment by jakehilton** [comment]
The inodes look great.

```
$ df -i
Filesystem      Inodes  IUsed   IFree IUse% Mounted on
udev           1923823    393 1923430    1% /dev
tmpfs          1925118    320 1924798    1% /run
/dev/xvda1      524288 107282  417006   21% /
none           1925118      2 1925116    1% /sys/fs/cgroup
none           1925118      1 1925117    1% /run/lock
none           1925118      4 1925114    1% /run/shm
none           1925118      2 1925116    1% /run/user
/dev/xvdb      5242880     65 5242815    1% /mnt
/dev/xvdc      5242880     11 5242869    1% /mnt2
```

**Comment by jakehilton** [comment]
This shows the device error slightly different.. referencing a job in a cache dir.

```
2016-01-04 04:02:07,952 [salt.loaded.int.returner.local_cache][WARNING ][17796] Could not write job cache file for minions: ['ausw3-encoder880']
2016-01-04 04:02:07,953 [salt.master      ][ERROR   ][17796] Error in function _return:
Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/salt/master.py", line 1379, in run_func
    ret = getattr(self, func)(load)
  File "/usr/lib/python2.7/dist-packages/salt/master.py", line 1192, in _return
    self.opts, load, event=self.event, mminion=self.mminion)
  File "/usr/lib/python2.7/dist-packages/salt/utils/job.py", line 88, in store_job
    mminion.returners[fstr](load)
  File "/usr/lib/python2.7/dist-packages/salt/returners/local_cache.py", line 137, in returner
    os.makedirs(hn_dir)
  File "/usr/lib/python2.7/os.py", line 157, in makedirs
    mkdir(name, mode)
OSError: [Errno 28] No space left on device: '/var/cache/salt/master/jobs/38/d5612cc4ab380d2b69b05321d21a49/ausw3-encoder880'
```

**Comment by cachedout** [comment]
System logs showing anything interesting? I'm surprised we're seeing an `OSError` here.

**Comment by jakehilton** [comment]
Nothing in the syslog on either the master or minion around that time.

**Comment by jakehilton** [comment]
I'd be happy to jump on a hangout if you'd like to have a bit more of an interactive look at it.

**Comment by cachedout** [comment]
I don't have time this morning I'm afraid, but shoot me an email and let's see if we can find some time.

**Comment by jakehilton** [comment]
I'm also seeing very high memory usage on the hung master. It's a dedicated box so nothing else is competing for resources. On the initial restart it's using just under 2 gig.. but now it's at 7.5 gig used.

**Comment by cachedout** [comment]
So, right now it looks like we've got at least three separate issues. I think we should split each of these errors (and the memory leak) into individual GH issues and track them independently.  Sound good?

**Comment by jakehilton** [comment]
I'm fine with that.. they all contribute to the master process hanging.

**Comment by jakehilton** [comment]
One other note.. I did see a server hang and the inodes were all used up. You had asked above so I thought I'd make note here.

**Comment by lukebeer** [comment]
Last night I issued "salt-cp '*xyz**abc*' src dst" and had the single master become unresponsive on all services, kicking me off ssh, irresponsive to icmp ping. <20 minions. Haven't been able to get on the server for logs yet.

**Comment by DmitryKuzmenko** [comment]
I've reproduced something looking close to this:
1. Setup multimaster environment with 2 masters and 1 minion.
2. Setup shceduler: alive every 30 seconds, state.apply every 20 seconds
3. Run all and check `test.ping` works from both masters.
4. Delete the minion key on all masters with `salt-key -y -D`
5. Run minion again.

After that either one or both masters don't respond to minion request and salt commands.

There are no any exception or error in master logs.

I'm continuing to investigate this.

**Comment by DmitryKuzmenko** [comment]
Caught it!

The problem I've found: if a minion's key was removed and the minion still think it's connected and perform interactions with the master, master handles requests and tries to respond but throws IOException in worker when trying to encrypt a response for minion because there's no minion key anymore. The exception doesn't appear in the log that is the reason it looked so magic.

Why it hangs: master worker uses zmq router-dealer patter with multiple REP workers connected to the dealer. When the exception occurs the worker didn't sent a reply that means REP socket will not handle any incoming messages.

Solution (not implemented yet): surely first we have to handle the exception. Also I think good idea would be to send some signal to the minion that would force it's reconnection to the failed master.

**Comment by jakehilton** [comment]
So would this manifest with the TCP transport as well? It doesn't use the zmq libraries correct? I don't seem to experience it when using the TCP transport hence my question.

We run a `salt-run manage.down removekeys=True` command periodically from the masters so if when they are run the minion doesn't respond they are removed.. and then if the minion still thinks they are connected it could fall into the scenario you explained if using zmq transport.

Great to see progress on this! Thanks for your help.

**Comment by DmitryKuzmenko** [comment]
@jakehilton for TCP transport there is the same case but it wouldn't fall to the same issue, i.e. master will not hang because TCP transport doesn't follow REP pattern in the master worker. So it's not require each message to be replied. But TCP transport need to be updated too to properly handle the case.

Thank you for good catch @jakehilton!

**Comment by cachedout** [comment]
This has been resolved by #33333
